File: apps/atlas/lib/field/field_journal.dart
// ============================================================
// As Above, So Below. As Within, So Without.
// The Future Dictates the Past and the Past is Always Present.
// ============================================================

import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:atlas_core/atlas_core.dart';
import 'package:atlas_geo/atlas_geo.dart';
import 'package:atlas_location/atlas_location.dart';
import 'package:atlas_tactical/atlas_tactical.dart';
import 'package:flutter/foundation.dart';
import 'package:path_provider/path_provider.dart';

const int kFieldJournalVersion = 1;
const String kFieldJournalDir = 'field_data';
const String kFieldJournalFile = 'journal.json';

enum WaypointSource { mapSelected, gpsRecorded }

String waypointSourceName(WaypointSource source) {
  return switch (source) {
    WaypointSource.mapSelected => 'map_selected',
    WaypointSource.gpsRecorded => 'gps_recorded',
  };
}

WaypointSource? waypointSourceFromName(Object? raw) {
  return switch (raw) {
    'map_selected' => WaypointSource.mapSelected,
    'gps_recorded' => WaypointSource.gpsRecorded,
    _ => null,
  };
}

final class StoredWaypoint {
  const StoredWaypoint({
    required this.id,
    required this.latitude,
    required this.longitude,
    required this.createdAt,
    this.label = '',
    this.note = '',
    required this.source,
  });

  final String id;
  final double latitude;
  final double longitude;
  final int createdAt;
  final String label;
  final String note;
  final WaypointSource source;

  Map<String, Object?> toJson() {
    return {
      'id': id,
      'latitude': latitude,
      'longitude': longitude,
      'created_at': createdAt,
      'label': label,
      'note': note,
      'source': waypointSourceName(source),
    };
  }

  static StoredWaypoint? tryParse(Object? raw) {
    if (raw is! Map) return null;
    final id = raw['id'];
    if (id is! String || !AtlasIds.check(id).isValid) return null;
    final latitude = (raw['latitude'] as num?)?.toDouble();
    final longitude = (raw['longitude'] as num?)?.toDouble();
    if (latitude == null || longitude == null) return null;
    if (!AtlasCoordinates.validate(latitude, longitude).isValid) return null;
    final createdAt = raw['created_at'];
    if (createdAt is! int) return null;
    final label = raw['label'];
    final note = raw['note'];
    final source = waypointSourceFromName(raw['source']);
    if (source == null) return null;
    return StoredWaypoint(
      id: id,
      latitude: latitude,
      longitude: longitude,
      createdAt: createdAt,
      label: label is String ? label : '',
      note: note is String ? note : '',
      source: source,
    );
  }
}

final class StoredTrack {
  const StoredTrack({
    required this.id,
    required this.createdAt,
    this.points = const [],
    required this.source,
  });

  final String id;
  final int createdAt;
  final List<StoredWaypoint> points;
  final WaypointSource source;

  int get pointCount => points.length;

  AtlasTrack toTrack() {
    return AtlasTrack(
      id: AtlasId(id),
      points: [
        for (final point in points)
          AtlasWaypoint(
            id: AtlasId(point.id),
            position: AtlasCoordinate(
              latitude: point.latitude,
              longitude: point.longitude,
            ),
            createdAt: point.createdAt,
            label: point.label,
            note: point.note,
          ),
      ],
      createdAt: createdAt,
    );
  }

  Map<String, Object?> toJson() {
    return {
      'id': id,
      'created_at': createdAt,
      'source': waypointSourceName(source),
      'points': [for (final point in points) point.toJson()],
    };
  }

  static StoredTrack? tryParse(Object? raw) {
    if (raw is! Map) return null;
    final id = raw['id'];
    if (id is! String || !AtlasIds.check(id).isValid) return null;
    final createdAt = raw['created_at'];
    if (createdAt is! int) return null;
    final source = waypointSourceFromName(raw['source']);
    if (source == null) return null;
    final rawPoints = raw['points'];
    if (rawPoints is! List) return null;
    final points = <StoredWaypoint>[];
    for (final entry in rawPoints) {
      final point = StoredWaypoint.tryParse(entry);
      if (point == null) return null;
      points.add(point);
    }
    return StoredTrack(
      id: id,
      createdAt: createdAt,
      points: points,
      source: source,
    );
  }
}

final class FieldJournal extends ChangeNotifier {
  FieldJournal({
    Future<Directory> Function()? directoryProvider,
    int Function()? clock,
  })  : _directoryProvider = directoryProvider ?? getApplicationDocumentsDirectory,
        _clock = clock ?? (() => DateTime.now().millisecondsSinceEpoch);

  final Future<Directory> Function() _directoryProvider;
  final int Function() _clock;
  final Map<String, StoredWaypoint> _waypoints = {};
  final Map<String, StoredTrack> _tracks = {};
  int _sequence = 0;
  int _trackSequence = 0;
  Object? _lastError;
  int _skipped = 0;

  List<StoredWaypoint> get waypoints => _waypoints.values.toList();
  StoredWaypoint? lookup(String id) => _waypoints[id];
  List<StoredTrack> get tracks => _tracks.values.toList();
  StoredTrack? lookupTrack(String id) => _tracks[id];
  Object? get lastError => _lastError;
  int get skippedCount => _skipped;

  StoredWaypoint create({
    required double latitude,
    required double longitude,
    String label = '',
    String note = '',
    WaypointSource source = WaypointSource.mapSelected,
  }) {
    final point = AtlasCoordinates.checked(latitude, longitude);
    final id = _allocateId();
    final record = StoredWaypoint(
      id: id,
      latitude: point.latitude,
      longitude: point.longitude,
      createdAt: _clock(),
      label: label,
      note: note,
      source: source,
    );
    _waypoints[id] = record;
    _lastError = null;
    unawaited(persist());
    notifyListeners();
    return record;
  }

  bool updateLabel(String id, String label) {
    final current = _waypoints[id];
    if (current == null) return false;
    _waypoints[id] = StoredWaypoint(
      id: current.id,
      latitude: current.latitude,
      longitude: current.longitude,
      createdAt: current.createdAt,
      label: label,
      note: current.note,
      source: current.source,
    );
    _lastError = null;
    unawaited(persist());
    notifyListeners();
    return true;
  }

  bool updateNote(String id, String note) {
    final current = _waypoints[id];
    if (current == null) return false;
    _waypoints[id] = StoredWaypoint(
      id: current.id,
      latitude: current.latitude,
      longitude: current.longitude,
      createdAt: current.createdAt,
      label: current.label,
      note: note,
      source: current.source,
    );
    _lastError = null;
    unawaited(persist());
    notifyListeners();
    return true;
  }

  bool remove(String id) {
    if (_waypoints.remove(id) == null) return false;
    _lastError = null;
    unawaited(persist());
    notifyListeners();
    return true;
  }

  StoredTrack? saveTrack({required List<AtlasLocationFix> fixes}) {
    if (fixes.isEmpty) return null;
    final id = _allocateTrackId();
    final points = <StoredWaypoint>[
      for (var i = 0; i < fixes.length; i++)
        StoredWaypoint(
          id: '$id-p${(i + 1).toString().padLeft(4, '0')}',
          latitude: fixes[i].position.latitude,
          longitude: fixes[i].position.longitude,
          createdAt: fixes[i].at,
          source: WaypointSource.gpsRecorded,
        ),
    ];
    final record = StoredTrack(
      id: id,
      createdAt: fixes.first.at,
      points: points,
      source: WaypointSource.gpsRecorded,
    );
    _tracks[id] = record;
    _lastError = null;
    unawaited(persist());
    notifyListeners();
    return record;
  }

  bool removeTrack(String id) {
    if (_tracks.remove(id) == null) return false;
    _lastError = null;
    unawaited(persist());
    notifyListeners();
    return true;
  }

  Future<void> persist() async {
    try {
      final dir = await _journalDir();
      await File('${dir.path}/$kFieldJournalFile')
          .writeAsString(jsonEncode(_envelope()), flush: true);
    } catch (error) {
      _lastError = error;
      notifyListeners();
    }
  }

  Future<void> restore() async {
    try {
      await _restoreUnsafe();
    } catch (error) {
      _lastError = error;
      notifyListeners();
    }
  }

  Map<String, Object?> _envelope() {
    return {
      'version': kFieldJournalVersion,
      'waypoints': [for (final record in _waypoints.values) record.toJson()],
      'tracks': [for (final record in _tracks.values) record.toJson()],
    };
  }

  Future<Directory> _journalDir() async {
    final base = await _directoryProvider();
    final dir = Directory('${base.path}/$kFieldJournalDir');
    if (!await dir.exists()) await dir.create(recursive: true);
    return dir;
  }

  Future<void> _restoreUnsafe() async {
    final dir = await _journalDir();
    final file = File('${dir.path}/$kFieldJournalFile');
    if (!await file.exists()) {
      _lastError = null;
      notifyListeners();
      return;
    }
    late final Object? decoded;
    try {
      decoded = jsonDecode(await file.readAsString());
    } catch (error) {
      _lastError = error;
      notifyListeners();
      return;
    }
    if (decoded is! Map || decoded['version'] != kFieldJournalVersion) {
      _lastError = StateError('unsupported field journal version');
      notifyListeners();
      return;
    }
    if (decoded['waypoints'] is! List) {
      _lastError = StateError('field journal has no waypoint list');
      notifyListeners();
      return;
    }
    final loaded = <String, StoredWaypoint>{};
    var skipped = 0;
    var maxSequence = 0;
    for (final raw in (decoded['waypoints'] as List)) {
      final record = StoredWaypoint.tryParse(raw);
      if (record == null || loaded.containsKey(record.id)) {
        skipped += 1;
        continue;
      }
      loaded[record.id] = record;
      final sequence = _sequenceOf(record.id);
      if (sequence > maxSequence) maxSequence = sequence;
    }
    final loadedTracks = <String, StoredTrack>{};
    var maxTrackSequence = 0;
    final rawTracks = decoded['tracks'];
    if (rawTracks != null) {
      if (rawTracks is! List) {
        _lastError = StateError('field journal has no track list');
        notifyListeners();
        return;
      }
      for (final raw in rawTracks) {
        final record = StoredTrack.tryParse(raw);
        if (record == null || loadedTracks.containsKey(record.id)) {
          skipped += 1;
          continue;
        }
        loadedTracks[record.id] = record;
        final sequence = _trackSequenceOf(record.id);
        if (sequence > maxTrackSequence) maxTrackSequence = sequence;
      }
    }
    _waypoints
      ..clear()
      ..addAll(loaded);
    _tracks
      ..clear()
      ..addAll(loadedTracks);
    _sequence = maxSequence;
    _trackSequence = maxTrackSequence;
    _skipped = skipped;
    _lastError = null;
    notifyListeners();
  }

  String _allocateId() {
    var candidate = _sequence + 1;
    while (_waypoints.containsKey(_idFor(candidate))) {
      candidate += 1;
    }
    _sequence = candidate;
    return _idFor(candidate);
  }

  String _allocateTrackId() {
    var candidate = _trackSequence + 1;
    while (_tracks.containsKey(_trackIdFor(candidate))) {
      candidate += 1;
    }
    _trackSequence = candidate;
    return _trackIdFor(candidate);
  }

  static String _idFor(int sequence) {
    return 'wp-${sequence.toString().padLeft(6, '0')}';
  }

  static int _sequenceOf(String id) {
    if (!id.startsWith('wp-')) return 0;
    return int.tryParse(id.substring(3)) ?? 0;
  }

  static String _trackIdFor(int sequence) {
    return 'trk-${sequence.toString().padLeft(6, '0')}';
  }

  static int _trackSequenceOf(String id) {
    if (!id.startsWith('trk-')) return 0;
    return int.tryParse(id.substring(4)) ?? 0;
  }
}

File: apps/atlas-android/app/src/main/java/com/sovereignatlas/atlas/field/FieldJournal.kt
// ============================================================
// As Above, So Below. As Within, So Without.
// The Future Dictates the Past and the Past is Always Present.
// ============================================================

package com.sovereignatlas.atlas.field

import com.sovereignatlas.atlas.core.AtlasContractException
import com.sovereignatlas.atlas.geo.AtlasCoordinates
import com.sovereignatlas.atlas.location.AtlasLocationFix
import java.io.File

class FieldJournal(
    private val directoryProvider: () -> File,
    private val clockMs: () -> Long = { System.currentTimeMillis() },
) {
    private val waypoints = LinkedHashMap<String, StoredWaypoint>()
    private val tracks = LinkedHashMap<String, StoredTrack>()
    private var sequence = 0
    private var trackSequence = 0
    private var skipped = 0
    private var lastError: Throwable? = null
    private val listeners = ArrayList<() -> Unit>()

    fun addListener(listener: () -> Unit) {
        listeners.add(listener)
    }

    fun removeListener(listener: () -> Unit) {
        listeners.remove(listener)
    }

    fun waypoints(): List<StoredWaypoint> = waypoints.values.toList()

    fun lookup(id: String): StoredWaypoint? = waypoints[id]

    fun tracks(): List<StoredTrack> = tracks.values.toList()

    fun lookupTrack(id: String): StoredTrack? = tracks[id]

    fun lastErrorOrNull(): Throwable? = lastError

    fun skippedCount(): Int = skipped

    fun create(
        latitude: Double,
        longitude: Double,
        label: String = "",
        note: String = "",
        source: WaypointSource = WaypointSource.mapSelected,
    ): StoredWaypoint {
        val point = AtlasCoordinates.checked(latitude, longitude)
        val id = allocateId()
        val record = StoredWaypoint(
            id = id,
            latitude = point.latitude,
            longitude = point.longitude,
            createdAt = clockMs(),
            label = label,
            note = note,
            source = source,
        )
        waypoints[id] = record
        lastError = null
        persist()
        notifyListeners()
        return record
    }

    fun updateLabel(id: String, label: String): Boolean {
        val current = waypoints[id] ?: return false
        waypoints[id] = current.copy(label = label)
        lastError = null
        persist()
        notifyListeners()
        return true
    }

    fun updateNote(id: String, note: String): Boolean {
        val current = waypoints[id] ?: return false
        waypoints[id] = current.copy(note = note)
        lastError = null
        persist()
        notifyListeners()
        return true
    }

    fun remove(id: String): Boolean {
        if (waypoints.remove(id) == null) return false
        lastError = null
        persist()
        notifyListeners()
        return true
    }

    fun saveTrack(fixes: List<AtlasLocationFix>): StoredTrack? {
        if (fixes.isEmpty()) return null
        val id = allocateTrackId()
        val points = fixes.mapIndexed { index, fix ->
            StoredWaypoint(
                id = "$id-p${(index + 1).toString().padStart(4, '0')}",
                latitude = fix.position.latitude,
                longitude = fix.position.longitude,
                createdAt = fix.atMs,
                source = WaypointSource.gpsRecorded,
            )
        }
        val record = StoredTrack(
            id = id,
            createdAt = fixes.first().atMs,
            points = points,
            source = WaypointSource.gpsRecorded,
        )
        tracks[id] = record
        lastError = null
        persist()
        notifyListeners()
        return record
    }

    fun removeTrack(id: String): Boolean {
        if (tracks.remove(id) == null) return false
        lastError = null
        persist()
        notifyListeners()
        return true
    }

    fun persist() {
        try {
            val dir = journalDir()
            File(dir, JournalCodec.JOURNAL_FILE).writeText(
                JournalJson.render(
                    JournalCodec.envelope(waypoints.values, tracks.values),
                ),
            )
        } catch (error: Throwable) {
            lastError = error
            notifyListeners()
        }
    }

    fun restore() {
        try {
            restoreUnsafe()
        } catch (error: Throwable) {
            lastError = error
            notifyListeners()
        }
    }

    private fun journalDir(): File {
        val dir = File(directoryProvider(), JournalCodec.JOURNAL_DIR)
        if (!dir.exists()) dir.mkdirs()
        return dir
    }

    private fun restoreUnsafe() {
        val file = File(journalDir(), JournalCodec.JOURNAL_FILE)
        if (!file.exists()) {
            lastError = null
            notifyListeners()
            return
        }
        val decoded = try {
            JournalJson.parse(file.readText())
        } catch (error: Throwable) {
            lastError = error
            notifyListeners()
            return
        }
        val data = try {
            JournalCodec.parseEnvelope(decoded)
        } catch (error: AtlasContractException) {
            lastError = IllegalStateException(error.rejection.message)
            notifyListeners()
            return
        }
        waypoints.clear()
        waypoints.putAll(data.waypoints)
        tracks.clear()
        tracks.putAll(data.tracks)
        sequence = data.waypointSequence
        trackSequence = data.trackSequence
        skipped = data.skipped
        lastError = null
        notifyListeners()
    }

    private fun allocateId(): String {
        var candidate = sequence + 1
        while (waypoints.containsKey(JournalCodec.waypointIdFor(candidate))) {
            candidate += 1
        }
        sequence = candidate
        return JournalCodec.waypointIdFor(candidate)
    }

    private fun allocateTrackId(): String {
        var candidate = trackSequence + 1
        while (tracks.containsKey(JournalCodec.trackIdFor(candidate))) {
            candidate += 1
        }
        trackSequence = candidate
        return JournalCodec.trackIdFor(candidate)
    }

    private fun notifyListeners() {
        for (listener in listeners.toList()) {
            listener()
        }
    }
}
